Universal media player

rmkraus opened the pull request:
Created the initial implementation of the Universal Media Player. This is not quite done, but I wanted to get feedback before I wrapped things up.

Currently, the configuration for this component looks like this:

``` yaml
media_player:
  platform: universal
  name: MEDIA PLAYER NAME
  children:
    - media_player.cast
    - media_player.roku
  commands:
    turn_on:
       service: SERVICE
       data: SERVICE_DATA
    turn_off: 
       service: SERVICE
       data: SERVICE_DATA
    volume_up: 
       service: SERVICE
       data: SERVICE_DATA
    volume_down: 
       service: SERVICE
       data: SERVICE_DATA
    volume_mute: 
       service: SERVICE
       data: SERVICE_DATA
  attributes:
    is_volume_muted: ENTITY_ID|ATTRIBUTE
    state: ENTITY_ID|ATTRIBUTE 
```

where `commands` and `attributes` are all optional.

At its core, it is just a collection of child media players. It will report the status from and operate the first one in the list that is active (not off and not idle). To allow better control of external devices, the volume and power commands can be overwritten. If turn_on and turn_off are overwritten, the state attribute can be overwritten in some instances. The following is the logic used to determine the state if it is "overwritten":

```
     off if master state is off                                                  
ELSE Status of first active child if there is an active child
ELSE master state if defined
ELSE off
```

This will still require an update to the frontend to add volume buttons with the support flag. I would also like to have this component update itself when any one of its children is updated. Other than that, this is basically it.

rmkraus committed:
Merge remote-tracking branch 'balloob/dev' into dev

rmkraus committed:
Merge remote-tracking branch 'balloob/dev' into dev

rmkraus committed:
Merge remote-tracking branch 'balloob/dev' into dev

rmkraus committed:
Added initial implementation of universal media player.

rmkraus committed:
Merge remote-tracking branch 'balloob/dev' into universal-media-player

rmkraus committed:
Added active_child attribute to universal media players.

The entity of the first active child is now reported in the attributes
for a universal media player.

rmkraus committed:
Better handling of entity lookups in Universal media player.

Allowed the lookup function in the Universal Media Player to return
either a state object or the actual state of an entity during lookup.

rmkraus commented:
Frontend support has been submitted with balloob/home-assistant-polymer#17

rmkraus committed:
Lint fixes and faster updating to universal media player.

1) Many lint fixes.
2) Bound the Universal Media Player to its dependencies so that its
state will be updated when one of its dependencies is changed.

rmkraus commented:
I think this component is just about finished. 

Once the frontend code is merged, I can update the frontend in this PR. I'll start adding some checks now.

balloob commented:
Can you write a few tests for this to ensure behavior is working as expected?

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+        if service_name in self._cmds:
+            cmd_data = self._cmds[service_name]
+            domain, service = cmd_data[CONF_SERVICE].split('.', 1)
+            self.hass.services.call(domain, service,
```
Please use the service helper [`call_from_config`](https://github.com/balloob/home-assistant/blob/dev/homeassistant/helpers/service.py#L14). It's the same code that powers calling services from config for the automation and alexa components.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+        active_child = self.active_child_state
+        service_data[ATTR_ENTITY_ID] = active_child.entity_id
+
+        self.hass.services.call(DOMAIN, service_name, service_data,
```
`call_from_config`

rmkraus replied:
This one isn't calling from a configuration. This one is forwarding the service call to the active child.

balloob replied:
You are correct. We should only use it for the configurable services.

anthonykeane commented:
Hi Guys,
I am interested in helping with this development, I have setup some Python scripts to control any samsung TV (power, channels, volume etc) and would love to be able to use "Universal Media Player" or "Universal Remote Control" platform.

Currently I use scripts to fire the controls, it works but is clunky.


I also have a LIRC server controlled by HA, again scripts. Not ideal.

Where can I assist?

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -72,6 +72,7 @@
 SUPPORT_TURN_ON = 128
 SUPPORT_TURN_OFF = 256
 SUPPORT_PLAY_MEDIA = 512
+SUPPORT_VOLUME_BUTTONS = 1024
```
I don't think volume buttons describes the functionality appropriately. The fact that they are represented as buttons is a UI concern, not a support flag concern. What about `SUPPORT_VOLUME_STEP` to indicate it supports modifying the volume in steps.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+
+        track_state_change(hass, self.dependencies, self.update_state)
+
+    def _entity_lkp(self, entity_id=None, state_attr=None, state_as_obj=True):
```
No need to make `entity_id` have a default value of `None`, that will not happen.

rmkraus replied:
This could happen if someone where to configure like this:

```yaml
...
  attributes:
    state:
```

I made it tolerate this without crashing, although it is an odd thing to do. I could change this if you'd prefer.

balloob replied:
Well, the fact that configuration can impact the number of arguments being sent to a method sounds like code smell :nose: Star args should be used sparingly, avoided if possible.

rmkraus replied:
This discussion got hidden a bit early... 

I agree 100%, this is fixed and it will more cleanly handle bad configurations.

rmkraus commented:
@anthonykeane If you could write a Samsung TV component, that would be helpful. It could possibly be a media player component that assumes states like mute and power. Alternatively, it could just create a set of switches. If you could expose power on command, power off command, assumed power state, mute toggle command, assumed is_muted state, volume up command, and volume down command in HA, that would be enough to plug into this component.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+    @property
+    def children(self):
+        """ Gets children and their current states """
+        return {child: self._entity_lkp(child) for child in self._children}
```
Just use `self.hass.states.get` here, that way you can drop the `state_as_obj` parameter from `_entity_lkp`.

balloob replied:
Actually, this is only used for `active_child_state`, no need to expose it as it's own property. Just embed it in that function.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+    @property
+    def active_child_state(self):
+        """ The state of the active child or None """
+        for child_state in self.children.values():
```
Please be explicit about returning `None` at the end of this function

balloob replied:
Actually, to speed this entity up a lot, please cache this value inside the `update` method.

rmkraus committed:
Added tests to universal media player and fixed bug

1) Fixed universal media player to maintain specified child order when
checking for active child.
2) Added many tests to universal media player.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+
+        self._child_service(service_name, **service_data)
+
+    def _child_service(self, service_name, **service_data):
```
Please use a `dict` to pass around service data instead of magic keywords.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+        active_child = self.active_child_state
+        return active_child.attributes.get(attr_name) if active_child else None
+
+    def _override_or_child_service(self, service_name, **service_data):
```
Please use a `dict` to pass around service data instead of magic keywords.

balloob replied:
You can merge this method with `_child_service` and have a `allow_override=True` keyword param.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+
+    def play_youtube(self, media_id):
+        """ Plays a YouTube media. """
+        self._child_service(SERVICE_YOUTUBE_VIDEO, media_id=media_id)
```
For these types, please use the `ATTR_*` constants.

rmkraus replied:
I couldn't find ATTR constants and I think the cast component hardcoded those parameter names as well. Would you like me to correct this for cast and universal?

balloob replied:
My bad, you're right. No need to fix it.

On Mon, Jan 11, 2016 at 10:52 PM, Ryan Kraus <notifications@github.com>
wrote:

> In homeassistant/components/media_player/universal.py
> <https://github.com/balloob/home-assistant/pull/871#discussion_r49420473>:
>
> > +    def media_previous_track(self):
> > +        """ Send previous track command. """
> > +        self._child_service(SERVICE_MEDIA_PREVIOUS_TRACK)
> > +
> > +    def media_next_track(self):
> > +        """ Send next track command. """
> > +        self._child_service(SERVICE_MEDIA_NEXT_TRACK)
> > +
> > +    def media_seek(self, position):
> > +        """ Send seek command. """
> > +        data = {ATTR_MEDIA_SEEK_POSITION: position}
> > +        self._child_service(SERVICE_MEDIA_SEEK, **data)
> > +
> > +    def play_youtube(self, media_id):
> > +        """ Plays a YouTube media. """
> > +        self._child_service(SERVICE_YOUTUBE_VIDEO, media_id=media_id)
>
> I couldn't find ATTR constants and I think the cast component hardcoded
> those parameter names as well. Would you like me to correct this for cast
> and universal?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/871/files#r49420473>.
>



-- 
PaulusSchoutsen.nl
It's nice to be important but it's more important to be nice.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,433 @@
+
+    def update_state(self, *_):
+        """ event to trigger a state update in HA """
+        self.update_ha_state()
```
You can make this `self.update_ha_state(True)` once the `update` method caches the active child.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,435 @@
+        config[CONF_ATTRS][key] = val.split('|', 1)
+
+
+class UniversalMediaPlayer(MediaPlayerDevice):
```
You forgot `should_poll` to `return False`. State updates to this entity are driven by state changes to the child entities.

rmkraus replied:
Done now.

rmkraus committed:
Using call_from_config in Universal Media Player

Changed universal media player to use the call_from_config helper to
call services specified in the configuration file. This was done to
copy what is done in the Alexa and Automation components.

rmkraus committed:
Renamed SUPPORT_VOLUME_STEP flag in media_player

1) Renamed SUPPORT_VOLUME_BUTTONS to SUPPORT_VOLUME_STEP
2) Removed unused imports from tests.

rmkraus committed:
Streamlined child state lookups in universal media player

1) Removed children property because it was only being used by one
method.
2) Removed option to return state as object from _entity_lkp as it was
no longer needed.
3) Used hass.states.get to get entity state objects.
4) Revised test to remove children property.

rmkraus committed:
Made universal media player cache active player

Revised universal media player to cache the active player when updating
the state when any of the children change. Revised tests to accommodate
this change.

rmkraus committed:
Changed universal media player to keep service attrs in dict

Revised universal media player to keep service data in a dictionary
rather than passing it around as magic parameters.

rmkraus committed:
Merged service calling method in universal media player

rmkraus committed:
Set universal media player to force refresh when updating HA

rmkraus committed:
Added should_poll = False to universal media player

rmkraus commented:
Thanks for the comments, Paulus. I think I have them all integrated so far.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,437 @@
+
+    def update_state(self, *_):
+        """ event to trigger a state update in HA """
+        self._cache_active_child_state()
```
If you rename this to `update`, it will be called automatically by `update_ha_state` when you pass in `True` as the first param.

rmkraus replied:
The problem is that this function is the one calling update_ha_state. So it would just sit in a loop.

balloob replied:
I was talking about `self._cache_active_child_state()`.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,437 @@
+        self._child_state = None
+
+    @property
+    def active_child_state(self):
```
If this is only used internally, no need to make a property as there is already an instance attribute.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,437 @@
+        return False
+
+    @property
+    def dependencies(self):
```
If only used once, no need to extract it in a function.

anthonykeane commented:
Thanks Ryan,

Will do, ( feeling my way around Git, so how do I deliver the code to the
team?, sorry if this is a dumb question.)



Anthony Keane
+61 407 278 230
<http://plus.google.com/112282442897041393312>
<http://au.linkedin.com/in/anthonykeane>

On Tue, Jan 12, 2016 at 4:40 PM, Ryan Kraus <notifications@github.com>
wrote:

> @anthonykeane <https://github.com/anthonykeane> If you could write a
> Samsung TV component, that would be helpful. It could possibly be a media
> player component that assumes states like mute and power. Alternatively, it
> could just create a set of switches. If you could expose power on command,
> power off command, assumed power state, mute toggle command, assumed
> is_muted state, volume up command, and volume down command in HA, that
> would be enough to plug into this component.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/871#issuecomment-170801447>
> .
>

rmkraus commented:
@anthonykeane 
https://home-assistant.io/developers/#submitting-improvements

If you have any questions, feel free to ask them in the Gitter chat room.

anthonykeane commented:
Sweet Thanks



Anthony Keane
+61 407 278 230
<http://plus.google.com/112282442897041393312>
<http://au.linkedin.com/in/anthonykeane>

On Wed, Jan 13, 2016 at 12:49 AM, Ryan Kraus <notifications@github.com>
wrote:

> @anthonykeane <https://github.com/anthonykeane>
> https://home-assistant.io/developers/#submitting-improvements
>
> If you have any questions, feel free to ask them in the Gitter chat room.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/871#issuecomment-170917029>
> .
>

maddox commented:
I read what this new component can DO, but could you provide a real world example of what it does and what it solved? I'm super interested.

Thanks!

balloob commented:
Example, I am running Android TV and it supports Chromecast and Plex. (and Plex over Chromecast) 

Right now Home Assistant will pick up my Chromecast and my Plex and present it as two media players. This media player would wrap those two media players and provide an entity called TV.

If you're using HDMI-CEC you could also turn the TV on/off, all via 1 media player.

I consider this component experimental and time will tell us if we made the right choices or what to tweak.

Bart274 commented:
That would be a nice addition! :)

> Op 12 jan. 2016 om 22:03 heeft Paulus Schoutsen <notifications@github.com> het volgende geschreven:
> 
> Example, I am running Android TV and it supports Chromecast and Plex. (and Plex over Chromecast)
> 
> Right now Home Assistant will pick up my Chromecast and my Plex and present it as two media players. This media player would wrap those two media players and provide an entity called TV.
> 
> If you're using HDMI-CEC you could also turn the TV on/off, all via 1 media player.
> 
> I consider this component experimental and time will tell us if we made the right choices or what to tweak.
> 
> —
> Reply to this email directly or view it on GitHub.
>

rmkraus commented:
In my house, I am using Insteon IRLincs to control my media devices. This not only allows me to wrap my media centers together, but also redirect power commands to my TV and volume commands to my receiver.

The main benefit of this component is it allows you to have one entity on the frontend to control an entire entertainment system.

rmkraus commented:
> I consider this component experimental and time will tell us if we made the right choices or what to tweak.

Oh Paulus, you're always so skeptical of my hairbrained ideas. :stuck_out_tongue_winking_eye:

rmkraus committed:
Removed property from universal media player

The active_child_state property was unnecessary as it was not being
referenced outside the class. This commit removes it and updates the
tests accordingly.

rmkraus committed:
Removed dependencies property from universal media player

The dependencies property was only being called once by the __init__
method so it was removed and the code was moved to the __init__ method.
The tests were updated to reflect this.

rmkraus committed:
Fixed attribute configuration handling in universal media player

Forced all parsed attribute configurations to be of length 2. Removed
entity_id=None option from entity lookups. Explicitly passed entity
lookup information to _entity_lkp.

balloob commented:
> > I consider this component experimental and time will tell us if we made the right choices or what to tweak.
>
> Oh Paulus, you're always so skeptical of my hairbrained ideas. :stuck_out_tongue_winking_eye: 

Just trying to prevent this :+1: 

![giphy](https://cloud.githubusercontent.com/assets/1444314/12283136/dd438628-b958-11e5-9e97-1789115974d3.gif)

balloob commented:
You can ignore the flake8 fixes. I'll get a new PR ready to fix those.

rmkraus committed:
Renamed update_state to update in universal media player

Renamed update_state method in universal media player to update so that
it would be called by HA when the state was being published. Moved the
update_ha_state to a function inside of __init__. Updated the tests
accordingly.

rmkraus commented:
I want a water jet pack! lol

I went through all your comments again to make sure I hit them all. Found a couple that GitHub hid before I could fix them. We should be alright now. I'll work on the frontend PR so we can merge that first.

rmkraus committed:
Updated frontend to newest commit

rmkraus commented:
The Future Starts Now!

This should be ready to merge pending any further comments.

balloob commented on homeassistant/components/media_player/universal.py:
```diff
@@ -0,0 +1,438 @@
+        else:
+            return None
+
+    def _cache_active_child_state(self):
```
This is not being used right now.

balloob commented:
Merged! a0ddda4bc6ba1c5cb5f9563824ffeb4d0b3c6087
